torch/_inductor/memory.py

```diff
@@ -0,0 +1,525 @@
+    A bfs-based greedy topological order.
+    The algorithm maintain the max memory so far.
+    At every iteration, for each scheduleable node, it computes:
+        - how much memory needs to be allocated for the output buffers of this node;
+        - how much memory can be freed as a result of executing this node.
+    This gives us two values for each node:
+        (1) mem1: memory during the execution of the node;
+        (2) mem2: memory after executing the node, after some input buffers are freed.
+    The greedy approach select as follows:
+        (i) if there are nodes whose mem1 values are below the max memory so far,
+            then pick the node with the lowest mem2 value;
+        (ii) otherwise, pick the one with the lowest mem1 value.
+    """
+
+    # compute nodes' number of unmet dependencies (for schedulability)
+    # initialize the list of nodes ready to be scheduled
+    nodes_to_schedule: Set[BaseSchedulerNode] = set()
+    for node in nodes:
+        # note that .unmet_dependencies could have deps with the same name
+        # and in that case, it should only be counted once
+        node.mpi.indegree = len(node.mpi.pred_nodes)
+        if node.mpi.indegree == 0:
+            nodes_to_schedule.add(node)
+
+    # compute buffers' number of unmet successors (used to decide when to free)
+    for buf in list(name_to_buf.values()) + list(name_to_input_buf.values()):
+        buf.mpi.outdegree = len(buf.mpi.succ_nodes)
+        if buf.get_name() in graph_outputs:
+            buf.mpi.outdegree += 1
+
+    # initialize memory estimations
+    live_memory = sum(
+        input_buf.mpi.size_free for input_buf in name_to_input_buf.values()
+    )
+
+    # this is the total output memory, which is a lower bound for peak memory
+    output_memory = sum(
+        name_to_buf[buf_name].mpi.size_free
+        for buf_name in graph_outputs
+        if buf_name in name_to_buf
+    )
+    max_memory = max(live_memory, output_memory)
+
+    # compute the amount of memory that is allocated when a node is scheduled
+    # and the amount of memory that can be freed when a node is scheduled
+    for i, node in enumerate(nodes):
+        node.mpi.index = i  # keep track of the original order
+        node.mpi.size = sum(buffer.mpi.size_alloc for buffer in node.get_outputs())
+        node.mpi.memory_to_free = 0
+        # 1. if a buffer read by this node is last used by this node
+        #    then the buffer can be freed
+        for buf in node.mpi.pred_buffers:
+            if buf.mpi.outdegree == 1:
+                node.mpi.memory_to_free += buf.mpi.size_free
+        # 2. if a buffer written by this node is used internally and
+        #    not needed afterwards, it can be freed
+        for buf in node.get_outputs():
+            if buf.mpi.outdegree == 0:
+                node.mpi.memory_to_free += buf.mpi.size_free
+
+    # schedule nodes one at a time
+    schedule: List[BaseSchedulerNode] = []
+    while nodes_to_schedule:
```

Comment:
I am slightly worried about the while loop here being an infinite loop if the graph is ill-constructed (it probably doesn't happen right now, but probably important to prevent it from happening since we might want to enable it by default). Wonder is there a way to add a constraint for "maximum number of iterations" (can be a function of the graph size), and then just exit from this algorithm if it's not solved within the maximum iterations?

Reply:
Added a check for the number of iterations for the while loops. If it is longer than expected, will raise a runtime error. Also added a "try - except" clause for each run in the main algorithm.